oracle/src/main/java/org/jdbi/v3/oracle/OracleReturning.java

```diff
@@ -56,83 +57,52 @@
  * }
  * </pre>
  * Though you can bind multiple params, and whatnot
+ * <p>
+ * This class is beta, and may be changed incompatibly or removed at any time.
  */
-@Deprecated
-public class OracleReturning<ResultType> implements StatementCustomizer
-{
+public class OracleReturning<ResultType> implements StatementCustomizer {
     private final RowMapper<ResultType> mapper;
     private final List<int[]> binds = new ArrayList<>();
     private StatementContext context;
     private List<ResultType> results;
-    private Class<?> oraclePS ;
-    private Method registerReturnParameter ;
-    private Method getReturnResultSet ;
-    private Object stmt ;
+    private OraclePreparedStatement stmt;
 
     /**
      * Provide a mapper which knows how to do positional access, sadly the
      * {@link BeanMapper} uses the names in the result set
      *
      * @param mapper Must use only positional access to the result set
      */
-    public OracleReturning(RowMapper<ResultType> mapper)
-    {
+    public OracleReturning(RowMapper<ResultType> mapper) {
         this.mapper = mapper;
-        try {
-            this.oraclePS = Class.forName("oracle.jdbc.OraclePreparedStatement");
-            this.registerReturnParameter = oraclePS.getMethod("registerReturnParameter", new Class[]{int.class, int.class});
-            this.getReturnResultSet = oraclePS.getMethod("getReturnResultSet");
-        }
-        catch (Exception e) {
-            throw new RuntimeException(e);
-        }
     }
 
     /**
      * @see StatementCustomizer#beforeExecution(java.sql.PreparedStatement, StatementContext)
      */
     @Override
-    public void beforeExecution(PreparedStatement stmt, StatementContext ctx) throws SQLException
-    {
+    public void beforeExecution(PreparedStatement stmt, StatementContext ctx) throws SQLException {
         this.context = ctx;
 
-        if (!oraclePS.isAssignableFrom(stmt.getClass())) {
-            try {
-                final Method get_delegate = stmt.getClass().getMethod("getDelegate");
-                final Object candidate = get_delegate.invoke(stmt);
-                if (!oraclePS.isAssignableFrom(candidate.getClass())) {
-                    throw new Exception("Obtained delegate, but it still wasn't an OraclePreparedStatement");
-                }
-                else {
-                    this.stmt = candidate ;
-                }
-            }
-            catch (Exception e) {
-                throw new IllegalStateException("Statement is not an OraclePreparedStatement, nor" +
-                                                "one which we know how to find it from", e);
-            }
-        }
-        else {
-            this.stmt = stmt ;
+        if (!stmt.isWrapperFor(OraclePreparedStatement.class)) {
+            throw new IllegalStateException("Statement is not an instance of, nor a wrapper of, OraclePreparedStatement");
         }
+        this.stmt = stmt.unwrap(OraclePreparedStatement.class);
```

Comment:
This use case for `OracleReturning`, and the request for supporting multiple result sets, has me wondering whether `StatementExecutor` ought to take a `PreparedStatement` (post execution) in lieu of the `RowMapper` and `ResultSet`:

```java
@FunctionalInterface
public interface StatementExecutor<R>
{
    R execute(PreparedStatement stmt, StatementContext ctx) throws SQLException;
}
```

Although I'm realizing now that `StatementExecutor` is a bit of a misnomer: the interface isn't executing the statement--it has already executed by this point.

Since `StatementExecutor` no longer receives the mapper, and `ResultBearing` has no access to it, then the current `ResultBearing.iterator()` method cannot be default--but we can just push it down to `Query`. `GeneratedKeys` (the only other implementation) already overrides `iterator()`.

This also suggests to me that we might want to extract a super-interface of `ResultBearing` that has the `execute(StatementExecutor)` and possibly `getContext()` methods. Thus `ResultBearing` would continue to operate for everything iterable (single result set), vs `StatementExecuting` (bleh) which only implies that the answer came from that `PreparedStatement` in some way.

This would allow us to bring other `SqlStatement` classes (`Update`, `Call`, probably not `Batch`) into the same fold as `Query`. We provide a default set of the most useful result strategies, and other folks can introduce their own strategies for esoteric scenarios.

It also allows third-party solutions for extracting results in esoteric scenarios, e.g. multiple result sets.